File: DNA Cryptography – Inspired by Biology.cpp
#include <iostream>
#include <string>
#include <bitset>
#include <ctime>
#include <cstdlib>

using namespace std;

// DNA Mapping
const string BINARY_TO_DNA[4] = {"A", "T", "C", "G"};
const string DNA_TO_BINARY = "00AT10CG11";

// Convert Text to Binary
string textToBinary(const string &text) {
    string binaryString = "";
    for (char c : text) {
        binaryString += bitset<8>(c).to_string();
    }
    return binaryString;
}

// Convert Binary to DNA Sequence
string binaryToDna(const string &binary) {
    string dna = "";
    for (size_t i = 0; i < binary.length(); i += 2) {
        int index = stoi(binary.substr(i, 2), nullptr, 2);
        dna += BINARY_TO_DNA[index];
    }
    return dna;
}

// Convert DNA to Binary Sequence
string dnaToBinary(const string &dna) {
    string binary = "";
    for (char nucleotide : dna) {
        size_t index = DNA_TO_BINARY.find(nucleotide);
        if (index != string::npos && index % 2 == 0) {
            binary += DNA_TO_BINARY.substr(index + 1, 2);
        }
    }
    return binary;
}

// Generate Random DNA Key
string generateDnaKey(int length) {
    string key = "";
    string nucleotides = "ATCG";
    srand(time(0));
    for (int i = 0; i < length; i++) {
        key += nucleotides[rand() % 4];
    }
    return key;
}

// XOR-Based DNA Encryption
string dnaXorEncrypt(const string &dna, const string &key) {
    string encrypted = "";
    for (size_t i = 0; i < dna.length(); i++) {
        int dnaBin = DNA_TO_BINARY.find(dna[i]) / 2;
        int keyBin = DNA_TO_BINARY.find(key[i]) / 2;
        encrypted += BINARY_TO_DNA[dnaBin ^ keyBin]; // XOR operation
    }
    return encrypted;
}

// XOR-Based DNA Decryption
string dnaXorDecrypt(const string &encryptedDna, const string &key) {
    string decrypted = "";
    for (size_t i = 0; i < encryptedDna.length(); i++) {
        int encBin = DNA_TO_BINARY.find(encryptedDna[i]) / 2;
        int keyBin = DNA_TO_BINARY.find(key[i]) / 2;
        decrypted += BINARY_TO_DNA[encBin ^ keyBin]; // Reverse XOR
    }
    return decrypted;
}

// Convert Binary to Text
string binaryToText(const string &binary) {
    string text = "";
    for (size_t i = 0; i < binary.length(); i += 8) {
        text += char(bitset<8>(binary.substr(i, 8)).to_ulong());
    }
    return text;
}

// Encrypt Message
void encryptMessage(const string &message, string &encryptedDna, string &key) {
    string binaryData = textToBinary(message);
    string dnaSequence = binaryToDna(binaryData);
    key = generateDnaKey(dnaSequence.length());
    encryptedDna = dnaXorEncrypt(dnaSequence, key);
}

// Decrypt Message
string decryptMessage(const string &encryptedDna, const string &key) {
    string decryptedDna = dnaXorDecrypt(encryptedDna, key);
    string binaryData = dnaToBinary(decryptedDna);
    return binaryToText(binaryData);
}

// Main Function
int main() {
    string message = "HELLO DNA";
    cout << "Original Message: " << message << endl;

    string encryptedDna, key;
    encryptMessage(message, encryptedDna, key);

    cout << "Encrypted DNA: " << encryptedDna << endl;
    cout << "DNA Key:       " << key << endl;

    string decryptedMessage = decryptMessage(encryptedDna, key);
    cout << "Decrypted Message: " << decryptedMessage << endl;

    return 0;
}

File: DNA Cryptography – Inspired by Biology.java
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

public class DNACryptography {
    // DNA Mapping Tables
    private static final Map<String, Character> BINARY_TO_DNA = new HashMap<>();
    private static final Map<Character, String> DNA_TO_BINARY = new HashMap<>();

    static {
        BINARY_TO_DNA.put("00", 'A');
        BINARY_TO_DNA.put("01", 'T');
        BINARY_TO_DNA.put("10", 'C');
        BINARY_TO_DNA.put("11", 'G');

        for (Map.Entry<String, Character> entry : BINARY_TO_DNA.entrySet()) {
            DNA_TO_BINARY.put(entry.getValue(), entry.getKey());
        }
    }

    // Convert Text to Binary
    public static String textToBinary(String text) {
        StringBuilder binaryString = new StringBuilder();
        for (char c : text.toCharArray()) {
            binaryString.append(String.format("%08d", Integer.parseInt(Integer.toBinaryString(c))));
        }
        return binaryString.toString();
    }

    // Convert Binary to DNA
    public static String binaryToDna(String binary) {
        StringBuilder dna = new StringBuilder();
        for (int i = 0; i < binary.length(); i += 2) {
            dna.append(BINARY_TO_DNA.get(binary.substring(i, i + 2)));
        }
        return dna.toString();
    }

    // Convert DNA to Binary
    public static String dnaToBinary(String dna) {
        StringBuilder binary = new StringBuilder();
        for (char nucleotide : dna.toCharArray()) {
            binary.append(DNA_TO_BINARY.get(nucleotide));
        }
        return binary.toString();
    }

    // Generate Random DNA Key
    public static String generateDnaKey(int length) {
        StringBuilder key = new StringBuilder();
        char[] nucleotides = { 'A', 'T', 'C', 'G' };
        Random random = new Random();
        for (int i = 0; i < length; i++) {
            key.append(nucleotides[random.nextInt(4)]);
        }
        return key.toString();
    }

    // XOR-Based DNA Encryption
    public static String dnaXorEncrypt(String dna, String key) {
        StringBuilder encrypted = new StringBuilder();
        for (int i = 0; i < dna.length(); i++) {
            int dnaBin = "ATCG".indexOf(dna.charAt(i));
            int keyBin = "ATCG".indexOf(key.charAt(i));
            encrypted.append("ATCG".charAt(dnaBin ^ keyBin)); // XOR operation
        }
        return encrypted.toString();
    }

    // XOR-Based DNA Decryption
    public static String dnaXorDecrypt(String encryptedDna, String key) {
        StringBuilder decrypted = new StringBuilder();
        for (int i = 0; i < encryptedDna.length(); i++) {
            int encBin = "ATCG".indexOf(encryptedDna.charAt(i));
            int keyBin = "ATCG".indexOf(key.charAt(i));
            decrypted.append("ATCG".charAt(encBin ^ keyBin)); // Reverse XOR
        }
        return decrypted.toString();
    }

    // Convert Binary to Text
    public static String binaryToText(String binary) {
        StringBuilder text = new StringBuilder();
        for (int i = 0; i < binary.length(); i += 8) {
            text.append((char) Integer.parseInt(binary.substring(i, i + 8), 2));
        }
        return text.toString();
    }

    // Encrypt a Message
    public static String[] encryptMessage(String message) {
        String binaryData = textToBinary(message);
        String dnaSequence = binaryToDna(binaryData);
        String dnaKey = generateDnaKey(dnaSequence.length());
        String encryptedDna = dnaXorEncrypt(dnaSequence, dnaKey);
        return new String[] { encryptedDna, dnaKey };
    }

    // Decrypt a Message
    public static String decryptMessage(String encryptedDna, String dnaKey) {
        String decryptedDna = dnaXorDecrypt(encryptedDna, dnaKey);
        String binaryData = dnaToBinary(decryptedDna);
        return binaryToText(binaryData);
    }

    // Main Function
    public static void main(String[] args) {
        String message = "HELLO DNA";
        System.out.println("Original Message: " + message);

        // Encrypt
        String[] encryptionResult = encryptMessage(message);
        String encryptedDna = encryptionResult[0];
        String dnaKey = encryptionResult[1];

        System.out.println("Encrypted DNA: " + encryptedDna);
        System.out.println("DNA Key:       " + dnaKey);

        // Decrypt
        String decryptedMessage = decryptMessage(encryptedDna, dnaKey);
        System.out.println("Decrypted Message: " + decryptedMessage);
    }
}
